### ISSUE TITLE
nnUNetTrainer Trainer issues with the latest version

### ISSUE THREAD

**Issue by kabbas570** [MIC-DKFZ/nnUNet#1643]
Hello,
I have recently updated the nnUNet-v2 to solve a bug  as suggested here, #1348 using 

git clone https://github.com/MIC-DKFZ/nnUNet.git
cd nnUNet
pip install -e .

I found that before this update I could change the learning rate and num_epochs from nnUNetTrainer , without specifying the variants, but now whatever I change in nnUNetTrainer , it runs for 1000 epochs and with its default learning rate.
1- Can you please suggest how to get back the previous behaviour? That is much easier and clean.
2- If not how we can use multiple of its variants simultaneously, say I want to train with no_deep supervision and also for 100 epochs, and wth my own learning rate startpoint.

Thank You
Cheers
Abbas

**Comment by Karol-G** [comment]
Hey @kabbas570,

The best way to modify any parameters in the trainer is to create a new trainer class in the `nnUNetTrainer` folder that inherits from `nnUNetTrainer`. Otherwise, things might break like in your case.

In your case this could be the code you want:

```python
import torch
from nnunetv2.training.nnUNetTrainer.nnUNetTrainer import nnUNetTrainer
from nnunetv2.utilities.label_handling.label_handling import determine_num_input_channels


class nnUNetTrainer_kabbas570(nnUNetTrainer):
    def __init__(self, plans: dict, configuration: str, fold: int, dataset_json: dict, unpack_dataset: bool = True, device: torch.device = torch.device('cuda')):
        super().__init__(plans, configuration, fold, dataset_json, unpack_dataset, device)

        self.initial_lr = 1e-2  # Change to something else
        self.num_epochs = 1000  # Change to something else
        self.initial_lr = 1e-2  # Change to something else
        self.enable_deep_supervision = True  # Change to something else

    def initialize(self):
        if not self.was_initialized:
            self.num_input_channels = determine_num_input_channels(self.plans_manager, self.configuration_manager,
                                                                   self.dataset_json)

            self.network = self.build_network_architecture(self.plans_manager, self.dataset_json,
                                                           self.configuration_manager,
                                                           self.num_input_channels,
                                                           enable_deep_supervision=self.enable_deep_supervision).to(self.device)
            # compile network for free speedup
            if ('nnUNet_compile' in os.environ.keys()) and (
                    os.environ['nnUNet_compile'].lower() in ('true', '1', 't')):
                self.print_to_log_file('Compiling network...')
                self.network = torch.compile(self.network)

            self.optimizer, self.lr_scheduler = self.configure_optimizers()
            # if ddp, wrap in DDP wrapper
            if self.is_ddp:
                self.network = torch.nn.SyncBatchNorm.convert_sync_batchnorm(self.network)
                self.network = DDP(self.network, device_ids=[self.local_rank])

            self.loss = self._build_loss()
            self.was_initialized = True
        else:
            raise RuntimeError("You have called self.initialize even though the trainer was already initialized. "
                               "That should not happen.")
```

You would then run it with:

`nnUNetv2_train DATASET_NAME_OR_ID UNET_CONFIGURATION FOLD -tr nnUNetTrainer_kabbas570`

Best,
Karol

**Comment by kabbas570** [comment]
@Karol-G thank you, got the idea, 
Cheers
Abbas
